### PR TITLE
Feat: Vault list reorder

### PR DESCRIPTION
**Pull request by aminsato**

### CODE REVIEW

File: app/src/main/java/com/vultisig/wallet/presenter/home/HomeViewModel.kt [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
     private val navigator: Navigator<Destination>
 ) : ViewModel() {
-    val vaults: MutableLiveData<List<Vault>> = MutableLiveData(listOf())
+
+    var vaults = MutableStateFlow<List<Vault>> (emptyList())
 
     init {
-        vaults.postValue(vaultDB.selectAll())
+        vaults.update {
+            vaultDB.selectAll()
```
Comment (YGolybev) [inline app/src/main/java/com/vultisig/wallet/presenter/home/HomeViewModel.kt:27]: As in [previous PR](https://github.com/vultisig/vultisig-android/pull/134#discussion_r1605977378), the update's lambda can be reevaluated more than once, so consider moving calculations out of it.

Comment (YGolybev) [reply app/src/main/java/com/vultisig/wallet/presenter/home/HomeViewModel.kt:27]: Also, if you don't care for the current value of state flow, using `Stateflow.value = value` may be a bit more optimal

**Review by YGolybev — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/vultisig/wallet/presenter/home/HomeViewModel.kt [hunk 2]
```diff
@@ -29,4 +38,17 @@ internal class HomeViewModel @Inject constructor(
...
+        val updatedPositionsList = vaults.value.toMutableList().apply {
+            add(to.index, removeAt(from.index))
+        }
+        vaults.update { updatedPositionsList }
+        reIndexJob?.cancel()
+        reIndexJob = viewModelScope.launch {
+            delay(500)
+            val indexedVaultList = vaultDB.updateVaultsFileIndex(updatedPositionsList)
+            vaults.update { indexedVaultList }
+        }
```
[the remark this answers is not in the document]
Comment (aminsato) [reply app/src/main/java/com/vultisig/wallet/presenter/home/HomeViewModel.kt:51]: onmove method invoked repeatedly per each position changing. so if the current position is 0 and it moved to position 2, onMove method called 2 times. so for preventing database update during item movement i cancel the job of reIndexing until the movement action finished and wait for 500ms and then reIndex the data.

File: app/src/main/java/com/vultisig/wallet/data/on_board/db/VaultDB.kt [hunk 3]
```diff
@@ -16,8 +18,27 @@ class VaultDB(private val context: Context) {
...
+        val file = vaultsFolder.resolve("${indexedVault.name}${FILE_POSTFIX}")
+        file.writeText(gson.toJson(indexedVault))
+    }
+
+
+    suspend fun updateVaultsFileIndex(vaultList: List<Vault>): List<Vault> = withContext(IO) {
+        vaultList.reversed().onEachIndexed { index, vault ->
+            if (vault.fileIndex != index)
+                upsertIndexed(index, vault)
+        }.reversed()
```
[the remark this answers is not in the document]
Comment (aminsato) [reply app/src/main/java/com/vultisig/wallet/data/on_board/db/VaultDB.kt:41]: Done

**Review by YGolybev — APPROVED** [review]

**Pull request merged**